addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/JSFunctionPassiveScanner.java

```diff
@@ -0,0 +1,205 @@
+                                        MESSAGE_PREFIX + "otherinfo", pattern, line));
+                        evidence.append("\n");
+                        break; // Only need to record this line once
+                    }
+                }
+            }
+        }
+        if (evidence.length() > 0) {
+            this.raiseAlert(msg, id, evidence.toString());
+        }
+    }
+
+    private void raiseAlert(HttpMessage msg, int id, String evidence) {
+        Alert alert = new Alert(getPluginId(), Alert.RISK_LOW, Alert.CONFIDENCE_LOW, getName());
+        alert.setDetail(
+                this.getDescription(),
+                msg.getRequestHeader().getURI().toString(),
+                "", // Param, not relevant for this vulnerability
+                "", // Attack, not relevant for passive vulnerabilities
+                "", // Other info not specified in message bundle
+                this.getSolution(),
+                this.getReference(),
+                evidence,
+                749, // CWE-749: Exposed Dangerous Method or Function
+                0, // WASC Id - return 0 if no relevant one
+                msg);
+
+        parent.raiseAlert(id, alert);
+    }
+
+    private static List<Pattern> getPatterns() {
+        if (patterns == null) {
+            patterns = new ArrayList<>();
+
+            try {
+                File f =
+                        new File(
+                                Constant.getZapHome()
+                                        + File.separator
+                                        + FUNC_LIST_DIR
+                                        + File.separator
+                                        + FUNC_LIST_FILE);
+                if (!f.exists()) {
+                    throw new IOException("Couldn't find resource: " + f.getAbsolutePath());
+                }
+                try (BufferedReader reader = new BufferedReader(new FileReader(f))) {
+                    String line;
+                    while ((line = reader.readLine()) != null) {
+                        line = line.trim();
+                        if (!line.startsWith("#") && line.length() > 0) {
+                            patterns.add(
+                                    Pattern.compile(
+                                            "\\b" + line + "\\b", Pattern.CASE_INSENSITIVE));
+                        }
+                    }
+                }
+            } catch (IOException e) {
+                LOGGER.error(
+                        "Error on opening/reading suspicious comments file: "
+                                + File.separator
+                                + FUNC_LIST_DIR
+                                + File.separator
+                                + FUNC_LIST_FILE
+                                + " Error: "
+                                + e.getMessage());
+            }
+            for (String payload : getJsFunctionPayloads().get()) {
```

Comment:
Ideally this should be handled outside this method, otherwise the scanner will miss changes done to the payloads once the `patterns` is initialised.

Reply:
Just wondering actually, how does the passive scan rule get notified when a new payload is uploaded? I didn't see an observer pattern anywhere and am largely basing my current payload implementation on what I could learn from other passive scan rules so I don't really understand it fully yet :")

Reply:
There's no notification, the provider always returns the latest payloads. It would be useful to tell that to cache the `Pattern`s, but that will require changes in the Custom Payloads add-on.

Reply:
Hmm would it be possible then at every scan to check whether the payload has changed? Is there any built in change detection in Custom Payloads? If not I can just do a manual check

Reply:
There isn't, it would have to be done here. But for the first iteration I'd not add the check, just create the `Pattern`s each time.